docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+a followup RPC call to B. A will keep ``UserRRef {100, 1}`` alive until it
+receives the acknowledge from B (**G2**, not the return value of the RPC call).
+This is necessary because A should not send out the delete message until all
+previous messages are received, otherwise, the ``OwnerRRef`` could be
+deleted before usage as we do not guarantee message delivery order. This is done
+by creating a child ``ForkId`` of RRef, holding them in a map until receives the
+owner confirms the child ``ForkId``. The figure below shows the message flow.
+
+.. image:: https://user-images.githubusercontent.com/16999635/69164845-b67e0e80-0abe-11ea-93fa-d24674e75a2b.png
+    :alt: user_to_owner_arg.png
+    :width: 500 px
+
+
+Note that the ``UserRRef`` could be deleted on B before func finishes or even
+starts. However this is OK, as at the time B sends out ACK for the child
+``ForkId``, it already acquired the ``OwnerRRef`` instance, which would prevent
+it been deleted too soon.
+
+
+Owner Share RRef with User
+--------------------------
+
+Owner to user is the simplest case, where the owner can update reference
+counting locally, and does not need any additional control message to notify
+others. Regarding **G2**, it is same as the parent receives the ACK from the
+owner immediately, as the parent is the owner.
+
+.. code::
+
+  import torch
+  import torch.distributed.rpc as RRef, rpc
+
+  # on worker B and worker C
+  def func(rref):
+    pass
+
+  # on worker B, creating a local RRef
+  rref = RRef("data")
+  # say the rref has RRefId 100
+  dist.rpc_async('C', func, args=(rref, ))
+
+
+.. image:: https://user-images.githubusercontent.com/16999635/69164921-c990de80-0abe-11ea-9250-d32ad00cf4ae.png
+    :alt: owner_to_user.png
+    :width: 500 px
+
+The figure above shows the message flow. Note that when the ``OwnerRRef`` exits
+scope after the rpc_async call, it will not be deleted, because internally
+there is a map to hold it alive if there is any known forks, in which case is
+``UserRRef {100, 2}``. (**G2**)
+
+
+User Share RRef with User
+-------------------------
+
+This is the most complicated case where caller user (parent ``UserRRef``),
+callee user (child ``UserRRef``), and the owner all need to get involved.
+
+.. code::
+
+  import torch
+  import torch.distributed.rpc as rpc
+
+  # on worker A and worker C
+  def func(rref):
+    pass
+
+  # on worker A
+  rref = rpc.remote('B', torch.add, args=(torch.ones(2), 1))
+  # say the rref has RRefId 100 and ForkId 1
+  rpc.rpc_async('C', func, args=(rref, ))
+
+.. image:: https://user-images.githubusercontent.com/16999635/69164971-d6adcd80-0abe-11ea-971d-6b7af131f0fd.png
+    :alt: user_to_user.png
+    :width: 500 px
+
+When C receives the child ``UserRRef`` from A, it sends out a fork request to
+the owner B. Later, when the B confirms the ``UserRRef`` on C, C will perform
+two actions in parallel: 1) send out the child ACK to A and 2) run the UDF.
```

Comment:
What's UDF? Considering linking to relevant doc for users who aren't familiar or add a short explanation.